8635 HOPWA CAPER ATC sheet

ttoomey opened the pull request:
## _Merging this PR_
- use the squash-merge strategy for PRs targeting a release-X branch

# Description
Adds Access to Care (ATC) reporting support to the FY2026 HOPWA CAPER.

- **Access to Care Sheet**: Implemented new reporting sheet to track housing stability outcomes, medical/income access, and supportive service intersections.
- **HOPWA CAPER Generator**: Added an enrichment step to the data pipeline to pull ATC outcomes from custom assessments using a historical lookback mechanism.
- **STRMU and Supportive Services Sheets**: Refactored household counting to be strictly service-based and deduplicated by HOH client ID to ensure consistency with expenditure reporting.
- **Site Configuration Admin**: Introduced database-backed application properties and an administrative UI to manage feature flags and dynamic field mappings for custom assessment data.
- **Reporting Framework**: Improved validation and error handling for question and column identifiers in the base question sheet builder.
- **Test Coverage**: Added specs for ATC outcome casting, service-based deduplication, and cross-sheet consistency.

### Type of Change
New feature

## Checklist before requesting review
[//]: # (Remove any items that are not applicable)
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] I have updated the documentation (or not applicable)
- [x] I have added spec tests (or not applicable)
- [x] I have provided testing instructions in this PR or the related issue (or not applicable)

ttoomey committed:
8635-hopwa-caper-atc

ttoomey committed:
Merge branch 'release-192' into 8635-hopwa-caper-atc

ttoomey committed:
wip

ttoomey committed:
wip

ttoomey committed:
wip

ttoomey committed:
8635-hopwa-caper-atc

ttoomey committed:
cleanup

ttoomey committed:
interface to manage app config properties

ttoomey committed:
fix

ttoomey committed:
fixes

ttoomey committed:
Merge branch 'release-192' into 8635-hopwa-caper-atc

ttoomey committed:
n+1, tidy

ttoomey committed:
small corrections

eanders commented on app/views/admin/app_config_properties/_action_menu.haml:
```diff
@@ -0,0 +1,7 @@
+  items = [
+    { link_to: { path: edit_admin_app_config_property_path(property) }, icon: :pencil, label: 'Edit' },
+  ]
+  items << { link_to: { path: admin_app_config_property_path(property), method: :delete, data: { confirm: "Delete '#{property.key}'?" } }, icon: :cross, label: 'Delete'}
```
There's no conditional here, so maybe just include both in the array definition?
```suggestion
  items = [
    { link_to: { path: edit_admin_app_config_property_path(property) }, icon: :pencil, label: 'Edit' },
    { link_to: { path: admin_app_config_property_path(property), method: :delete, data: { confirm: "Delete '#{property.key}'?" } }, icon: :cross, label: 'Delete'},
  ]
```

eanders commented on app/views/admin/configs/index.haml:
```diff
@@ -7,6 +7,7 @@
   .row
     .col-sm-3
       %nav.nav.flex-column
+        = link_to 'App Config Properties', admin_app_config_properties_path, class: 'nav-item nav-link'
```
This is probably fine, but I'd love to move all of these pages into the main menu now that we have nested menus.  No one can ever find Available File Tags or Public Files.

eanders commented on app/controllers/admin/app_config_properties_controller.rb:
Should we add a simple feature doc for App Config Properties?

eanders commented on drivers/hopwa_caper/app/models/hopwa_caper/generators/fy2026/generator.rb:
```diff
@@ -288,5 +294,91 @@ def custom_service_rows_for(enrollment_context, date_range:)
+          order(AssessmentDate: :desc, id: :desc).
+          preload(:enrollment).group_by { |r| r.enrollment.id }
+
+        batch.each do |spm_enrollment|
```
Why are we calling these `spm_enrollment` should it be `hopwa_caper_enrollment`?

eanders replied:
or just `report_enrollment`

eanders reviewed (commented):
@ttoomey let me know if you want to discuss any of my questions.  I think the app config change looks like a significant improvement, but had some concern about variable names.

ttoomey committed:
pr feedback

ttoomey committed:
bugfix, uni test

ttoomey committed:
table data format

ttoomey committed:
qa feedback

ttoomey committed:
strmu changes

ttoomey committed:
bugfixes, docs, and validation tests

ttoomey commented:
changes since last review:

Distinguishes explicit "No Income/Insurance" responses from missing data and tightens household counting logic for STRMU service-based eligibility.

- **Admin Config UI**: Strip whitespace from keys and values on save; display compact timestamps in config/compliance/content page lists, add a better datetime format for tabular data
- **HOPWA CAPER Enrollment Staging**: Track income and insurance types with explicit sentinel markers (`NoIncomeSource`, `NoInsuranceSource`) to separate households that answered "No" from those with no assessment data
- **Income/Insurance Filters**: Add filter IDs for lookups/convenience accessor 
- **Access to Care Custom Data Elements**: improve custom assessment boolean values nil-handling to distinguish blank responses from Yes/No answers; skip nil values when searching for most recent valid response
- **Access to Care Sheet**: Deduplicate household counts within activity types; count heads of household consistently across all rows; use filter helpers for income/insurance queries
- **STRMU Sheet**: consistently apply requirement for financial assistance services received during reporting period rather than enrollment dates alone


### Type of Change
Bug fix

ttoomey commented on app/views/admin/compliance_requirements/_list.haml:
```diff
@@ -19,5 +20,6 @@
           - else
             %span.text-muted Never
         %td= req.active? ? 'Active' : 'Inactive'
+        %td= l req.updated_at, format: :table_compact
```
small opportunistic change for consistency in recent admin UI

eanders replied:
This is probably ok to leave, though while `MM/DD/YYYY` is the US standard format for dates, historically this application has used `Jan 20, 2025`.  It's something we should probably revisit globally so we maintain consistency across the application for date and time formats.

ttoomey replied:
Yes, it is a change. I suggest that we standardize on this format for dates & times in a tabular context as it optimizes for scannability, has a consistent width, and takes up less horizontal real-estate

ttoomey commented on app/views/admin/content_pages/_list.haml:
```diff
@@ -18,5 +18,5 @@
           - else
             %span.text-muted None
-        %td= page.updated_at.to_fs(:short)
+        %td= l page.updated_at, format: :table_compact
```
small opportunistic change for consistency in recent admin UI

ttoomey commented on config/locales/en.yml:
```diff
@@ -83,6 +83,9 @@ en:
       search_prompt: 'Search projects...'
+  time:
+    formats:
+      table_compact: "%m/%d/%y %I:%M%P"
```
this is more compact and has consistent width which seems better for tabular views

eanders replied:
See: https://github.com/greenriver/hmis-warehouse/pull/6101/changes#r2649157535

ttoomey committed:
fixes and specs, improve exceptions

ttoomey committed:
uniqness fix

ttoomey committed:
Merge branch 'release-193' into 8635-hopwa-caper-atc

ttoomey commented on app/models/hud_reports/question_sheet.rb:
```diff
@@ -13,7 +13,7 @@ class QuestionSheet
     def initialize(report:, question:)
       @report = report
-      raise "invalid question id #{question.inspect}" unless question =~ /\AQ\d+/
+      raise ArgumentError, "invalid question id #{question.inspect}" unless question =~ /\AQ\d+/
```
added type to support unit tests

eanders replied:
Aren't SPM questions actually things like `Measure 1`?

eanders replied:
Oh, I see, this isn't a change to the check, just what is raised.  Looks ok.

ttoomey committed:
cleanup

ttoomey commented on spec/models/hud_reports/question_sheet_spec.rb:
```diff
@@ -0,0 +1,147 @@
+
+require 'rails_helper'
+
+RSpec.describe HudReports::QuestionSheet do
```
opportunistically adding a unit test for this core utility class

ttoomey commented on drivers/hopwa_caper/spec/models/generators/fy2026/generator_spec.rb:
```diff
@@ -0,0 +1,223 @@
+require 'rails_helper'
+require_relative 'hopwa_caper_shared_context'
+
+RSpec.describe HopwaCaper::Generators::Fy2026::Generator, type: :model do
```
add missing unit test for generator

ttoomey committed:
fix test

ttoomey committed:
app config prop fix

ttoomey commented:
Changes since last review:
- **Base Reporting Sheet**: Adds a `services_with_hoh` helper to join services with Head of Household enrollments and refactors scoping to support client-based unduplication.
- **Access to Care Sheet and Supportive Services Sheet**: Shifts household counts from household IDs to distinct Head of Household client IDs.
- **STRMU Sheet**: Refactors service categorization (exclusive vs. multi-type) to group by client ID, ensuring clients with multiple enrollments are correctly deduplicated.
- **Test Coverage**: Adds validation for clients with multiple enrollments and varied service types to ensure accurate categorization and deduplication.
- **App config properties**: handles json values

eanders reviewed (approved):
Thanks @ttoomey !

ttoomey committed:
fix: scope income universe to adults

ttoomey committed:
Merge branch 'release-193' into 8635-hopwa-caper-atc

ttoomey committed:
fix bug in report value display for arrays

ttoomey committed:
presenter rollup, specs

ttoomey commented on app/models/hud_reports/report_client_base.rb:
```diff
@@ -72,7 +72,7 @@ def fetch_cell_value(col)
       if value.is_a?(Array)
         # For Arrays, calculate each array element's value using the column name for the array
-        value.map { |item| display_value(col, pii_policy: pii_policy, include_content_tag: include_content_tag, cell_val: item, calculate_cell: false) }
+        value = value.map { |item| display_value(col, pii_policy: pii_policy, include_content_tag: include_content_tag, cell_val: item, calculate_cell: false) }
```
This appears to be a broader bug, causes values containing arrays to skip display value

Note, the report client base seems like kind of a lot so I opted to build a presenter for the hopwa caper instead

eanders replied:
Agreed, this looks like a bug since it doesn't return or save the results.

ttoomey commented on drivers/hopwa_caper/app/models/hopwa_caper/drilldown_presenter.rb:
```diff
@@ -0,0 +1,174 @@
+# frozen_string_literal: true
+
+module HopwaCaper
+  class DrilldownPresenter
```
extracts the drill-down logic from the hopwa caper enrollment; should be more maintainable

eanders replied:
This is probably fine, but it feels like much of this is generic, so in the future we should probably have a base DrilldownPresenter and then inherit from it and just override `enrollment_fields` and maybe have some mechanism to determine if a custom `display_value` is needed (like for `services_summary`).

ttoomey replied:
Definitely, maybe the next time we need to mess with drilldowns on another report we can extract the common pieces to a base class

ttoomey committed:
remove services from drilldown (unused)

ttoomey commented on drivers/hopwa_caper/app/controllers/hopwa_caper/cells_controller.rb:
```diff
@@ -18,10 +18,6 @@ def show
         preload(enrollment: :project).
         joins(hud_reports_universe_members: { report_cell: :report_instance }).
         merge(::HudReports::ReportCell.for_table(@table).for_cell(@cell))
-      @services = @report.hopwa_caper_services.
```
we aren't actually using services in a drilldown. Removed to simplify

ttoomey commented:
@eanders 
sorry for the additional review. Changes:
* fixed an issue in the base report client display value
* implement a presenter for the hopwa caper rather than the shared implementation which seems overly complex
* added support for "services summary" column to support data visibility into service records
* removed support for services from hopwa caper drilldowns (only enrollment are used)

ttoomey committed:
bugfix
